planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
+  both the stream and event data. This would potentially help for the async daemon to parallelize work. It would definitely help out for "live" projections by having 
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
+* Re-evaluate the `ViewProjection` fluent interface. That might fit in fine to our new model of projections for v4 with more overloads
+* I've been in favor of having adapters for existing projection libraries like [Liquid Projections](https://liquidprojections.net/) in the past, but I think we'd be losing too
+  much optization, especially for the aggregations inside the async daemon.
+
+#### Defining a Projection to a Document
+
+
+```
+ProjectTo<ProjectedDocument>()
+    // How to transform the events to the projected document, see below
+    .TransformWith<SampleTransformer>()
+
+    // Optionally aggregate by stream
+    .AggregateByStream() // Think this is self-descriptive
+
+    .AggregateByTenant() // Not sure how common this would be. Aggregate across the entire tenant
+
+    // Or aggregate by something special, hopefull use a common base type here for the event type
+    .AggregateByEventData<TEvent>(Expression<Func<TEvent, int>> expression) // what's the identity of the projected, aggregate doc?
```

Comment:
Would be nice to see the example usage.

Reply:
This would be doing some kind of aggregation across streams, like maybe
`.AggregateByEventData<NewResident>(x => x.PostalCode)`

Reply:
That sounds interesting but I'm not sure if the current ViewProjections api and selecting the projection id from the event is more straightforward. It would be nice to gather some feedback from our Event Store users about the preferences.

Reply:
I use `ViewProjection` a lot for projecting to different ids than the stream id. Honestly the most common usage of projections for me is wanting to project to a different id than the stream id. Not entirely sure which "is the best" here though tbh.

Reply:
Saw here. For the same stream, normally inline projections were enough in most of the cases.

Reply:
@wastaz Would that be as important if you had more ability to vary the stream identity? I was mostly thinking about aggregates across streams here.

Reply:
We could make it a lower priority item, but it's surely gonna come up sooner or later. Or maybe it's something you handle more easily through being able to project to flat tables, and let standard SQL do its thing

Reply:
I agree.